### PR TITLE
IGNITE-17146 Add support for alpha, beta, etc. releases in IgniteProductVersion

### PR DESCRIPTION
**Pull request by tkalkirill**
https://issues.apache.org/jira/browse/IGNITE-17146

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/properties/IgniteProductVersion.java [hunk 1]
```diff
@@ -29,11 +30,9 @@
  */
 public class IgniteProductVersion implements Serializable {
     private static final Pattern VERSION_PATTERN =
-            Pattern.compile("(?<major>\\d+)\\.(?<minor>\\d+)\\.(?<maintenance>\\d+)((?<snapshot>-SNAPSHOT)|-(?<alpha>alpha\\d+))?");
+            Pattern.compile("(?<major>\\d+)\\.(?<minor>\\d+)\\.(?<maintenance>\\d+)(\\.(?<patch>\\d+))?(-(?<preRelease>[0-9A-Za-z]+))?");
```
Comment (sashapolo) [inline modules/core/src/main/java/org/apache/ignite/internal/properties/IgniteProductVersion.java:33]: it's better to use non-capturing groups, e.g.: `(?:\\.(?<patch>\\d+))`

Comment (sashapolo) [inline modules/core/src/main/java/org/apache/ignite/internal/properties/IgniteProductVersion.java:33]: Can you please add a description about the format to the javadoc? E.g. what `preRelease` is and some examples for Ignite versions?

File: modules/core/src/main/java/org/apache/ignite/internal/properties/IgniteProductVersion.java [hunk 2]
```diff
@@ -109,41 +112,48 @@ public byte maintenance() {
     }
 
     /**
-     * Returns {@code true} if this is a snapshot release, {@code false} otherwise.
+     * Returns the patch version number.
```
Comment (sashapolo) [inline modules/core/src/main/java/org/apache/ignite/internal/properties/IgniteProductVersion.java:115]: Please add a description when this method can return `{@code null}`

File: modules/core/src/main/java/org/apache/ignite/internal/properties/IgniteProductVersion.java [hunk 3]
```diff
@@ -109,41 +112,48 @@ public byte maintenance() {
...
      */
-    public boolean snapshot() {
-        return isSnapshot;
+    public @Nullable Byte patch() {
+        return patch;
     }
 
     /**
-     * Returns the alpha version of this release or an empty string if this is not an alpha release.
+     * Returns the pre-release version.
```
Comment (sashapolo) [inline modules/core/src/main/java/org/apache/ignite/internal/properties/IgniteProductVersion.java:122]: Same here

File: modules/core/src/main/java/org/apache/ignite/internal/properties/IgniteProductVersion.java (hunk 1 above)
Comment (sashapolo) [reply modules/core/src/main/java/org/apache/ignite/internal/properties/IgniteProductVersion.java:33]: Can we replace `[0-9A-Za-z]` with `\w`?

Comment (tkalkirill) [reply modules/core/src/main/java/org/apache/ignite/internal/properties/IgniteProductVersion.java:33]: Why is it better?

Comment (tkalkirill) [reply modules/core/src/main/java/org/apache/ignite/internal/properties/IgniteProductVersion.java:33]: From [docs](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/regex/Pattern.html) \w - A word character: [a-zA-Z_0-9] which doesn't quite fit

Comment (sashapolo) [reply modules/core/src/main/java/org/apache/ignite/internal/properties/IgniteProductVersion.java:33]: why? Are you worried about the underscore?

Comment (sashapolo) [reply modules/core/src/main/java/org/apache/ignite/internal/properties/IgniteProductVersion.java:33]: It's a minor optimization to save a little bit of memory

Comment (tkalkirill) [reply modules/core/src/main/java/org/apache/ignite/internal/properties/IgniteProductVersion.java:33]: I tried to fix it.

File: modules/rest-api/src/main/java/org/apache/ignite/internal/rest/api/cluster/IgniteProductVersionDto.java [hunk 4]
```diff
@@ -38,26 +39,28 @@ public class IgniteProductVersionDto {
     /** Maintenance version number. */
     private final short maintenance;
 
-    /** Flag indicating if this is a snapshot release. */
-    private final boolean isSnapshot;
+    /** Patch version number. */
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/rest-api/src/main/java/org/apache/ignite/internal/rest/api/cluster/IgniteProductVersionDto.java:42]: Fix it

File: modules/core/src/main/java/org/apache/ignite/internal/properties/IgniteProductVersion.java (hunk 1 above)
Comment (tkalkirill) [reply modules/core/src/main/java/org/apache/ignite/internal/properties/IgniteProductVersion.java:33]: Fixed it, I think

Comment (tkalkirill) [reply modules/core/src/main/java/org/apache/ignite/internal/properties/IgniteProductVersion.java:33]: According to the [SemVer2](https://semver.org/#spec-item-9) pre release cannot contain underscores.

File: modules/core/src/main/java/org/apache/ignite/internal/properties/IgniteProductVersion.java (hunk 2 above)
Comment (tkalkirill) [reply modules/core/src/main/java/org/apache/ignite/internal/properties/IgniteProductVersion.java:115]: Will it be clearer? `{@code null} if no patch version has been specified.`

File: modules/core/src/main/java/org/apache/ignite/internal/properties/IgniteProductVersion.java (hunk 3 above)
Comment (tkalkirill) [reply modules/core/src/main/java/org/apache/ignite/internal/properties/IgniteProductVersion.java:122]: Will it be clearer? `{@code null} if no pre-release version has been specified.`

File: modules/core/src/main/java/org/apache/ignite/internal/properties/IgniteProductVersion.java (hunk 1 above)
Comment (sashapolo) [reply modules/core/src/main/java/org/apache/ignite/internal/properties/IgniteProductVersion.java]: Looks awesome!

File: modules/core/src/main/java/org/apache/ignite/internal/properties/IgniteProductVersion.java (hunk 2 above)
Comment (sashapolo) [reply modules/core/src/main/java/org/apache/ignite/internal/properties/IgniteProductVersion.java:116]: yes

File: modules/core/src/main/java/org/apache/ignite/internal/properties/IgniteProductVersion.java (hunk 3 above)
Comment (sashapolo) [reply modules/core/src/main/java/org/apache/ignite/internal/properties/IgniteProductVersion.java:123]: yes

File: modules/core/src/main/java/org/apache/ignite/internal/properties/IgniteProductVersion.java (hunk 1 above)
Comment (tkalkirill) [reply modules/core/src/main/java/org/apache/ignite/internal/properties/IgniteProductVersion.java]: Fix it

File: modules/core/src/main/java/org/apache/ignite/internal/properties/IgniteProductVersion.java (hunk 2 above)
Comment (tkalkirill) [reply modules/core/src/main/java/org/apache/ignite/internal/properties/IgniteProductVersion.java:116]: Fix it

File: modules/core/src/main/java/org/apache/ignite/internal/properties/IgniteProductVersion.java (hunk 3 above)
Comment (tkalkirill) [reply modules/core/src/main/java/org/apache/ignite/internal/properties/IgniteProductVersion.java]: Fix it

**Review by sashapolo — APPROVED** [review]

**Pull request merged**